applications/MeshingApplication/custom_utilities/tet10_refinement_utility.h

```diff
@@ -0,0 +1,499 @@
+    * @param New_Elements: The new elements created
+    * @param interpolate_internal_variables: A boolean that defines if it is necessary to interpolate the internal variables
+    * @return this_model_part: The model part of the model (it is the input too)
+    */
+
+    void EraseOldElementAndCreateNewElement(
+            ModelPart& this_model_part,
+            const compressed_matrix<int>& Coord,
+            PointerVector< Element >& NewElements,
+            bool interpolate_internal_variables
+    ) override
+    {
+	ElementsArrayType& rElements = this_model_part.Elements();
+	ElementsArrayType::iterator it_begin = rElements.ptr_begin();
+	ElementsArrayType::iterator it_end = rElements.ptr_end();
+	unsigned int to_be_deleted = 0;
+	unsigned int large_id = (rElements.end() - 1)->Id() * 15;
+	unsigned int current_id = (rElements.end() - 1)->Id() + 1;
+	bool create_element = false;
+	int number_elem = 0;
+	int splitted_edges = 0;
+	int internal_node = 0;
+
+	const ProcessInfo& rCurrentProcessInfo = this_model_part.GetProcessInfo();
+	int edge_ids[6];
+	int t[56];
+	std::vector<int> aux;
+
+        std::cout << "****************** REFINING MESH ******************" << std::endl;
+        std::cout << "OLD NUMBER ELEMENTS: " << rElements.size() << std::endl;
+
+	for (int & i : t)
+	{
+	    i = -1;
+	}
+
+	for (ElementsArrayType::iterator it = it_begin; it != it_end; ++it)
+	{
+	  Element::GeometryType& geom = it->GetGeometry();
+          CalculateEdges(geom, Coord, edge_ids, aux);
+
+	  create_element = TetrahedraSplit::Split_Tetrahedra(edge_ids, t, &number_elem, &splitted_edges, &internal_node);
+
+	  if (create_element == true)
+	  {
+	      to_be_deleted++;
+
+          GlobalPointersVector< Element >& rChildElements = it->GetValue(NEIGHBOUR_ELEMENTS);
+          // We will use this flag to identify the element later, when operating on
+          // SubModelParts. Note that fully refined elements already have this flag set
+          // to true, but this is not the case for partially refined element, so we set it here.
+          it->SetValue(SPLIT_ELEMENT,true);
+          rChildElements.resize(0);
+	      
+		  unsigned int i0 = aux[0];
+		  unsigned int i1 = aux[1];
+		  unsigned int i2 = aux[2];
+		  unsigned int i3 = aux[3];
+          unsigned int i4 = aux[4];
+		  unsigned int i6 = aux[5];
+		  unsigned int i7 = aux[6];
+		  unsigned int i5 = aux[7];
+          unsigned int i8 = aux[8];
+		  unsigned int i9 = aux[9];
+          
+
+		  Tetrahedra3D10<Node < 3 > > geom(
+		      this_model_part.Nodes()(i0),
+		      this_model_part.Nodes()(i1),
+		      this_model_part.Nodes()(i2),
+		      this_model_part.Nodes()(i3),
+              this_model_part.Nodes()(i4),
+		      this_model_part.Nodes()(i5),
+		      this_model_part.Nodes()(i6),
+		      this_model_part.Nodes()(i7),
+              this_model_part.Nodes()(i8),
+		      this_model_part.Nodes()(i9)
+		  );
+
+		  // Generate new element by cloning the base one
```

Comment:
This part would be better to be done a separate method.

Reply:
Done :)